File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/ipc/WebMessageEndpoint.kt
package org.dweb_browser.core.ipc

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.dweb_browser.core.ipc.helper.DWebMessage
import org.dweb_browser.core.ipc.helper.EndpointProtocol
import org.dweb_browser.core.ipc.helper.IWebMessagePort
import org.dweb_browser.core.ipc.helper.cborToEndpointMessage
import org.dweb_browser.core.ipc.helper.jsonToEndpointMessage

open class WebMessageEndpoint(
  override val debugId: String,
  parentScope: CoroutineScope,
  val port: IWebMessagePort,
) : CommonEndpoint(parentScope) {

  override fun toString() = "WebMessageEndpoint@$debugId"

  override suspend fun postTextMessage(data: String) {
    port.postMessage(DWebMessage.DWebMessageString(data))
  }

  override suspend fun postBinaryMessage(data: ByteArray) {
    port.postMessage(DWebMessage.DWebMessageBytes(data))
  }

  override suspend fun doStart() {
    super.doStart()
    scope.launch {
      for (event in port.onMessage) {
        runCatching {
          val packMessage =
            if (protocol == EndpointProtocol.CBOR && event is DWebMessage.DWebMessageBytes) {
              cborToEndpointMessage(event.binary)
            } else {
              jsonToEndpointMessage(event.text)
            }
          endpointMsgChannel.send(packMessage)
        }.getOrElse { error ->
          debugEndpoint("WebEndpointOnMessage", event, error)
        }
      }
    }
  }

  init {
    beforeClose = {
      port.close(it)
    }
  }
}

File: desktop-dev/src/core/ipc/endpoint/WebMessageEndpoint.ts
import { $cborToEndpointMessage, $jsonToEndpointMessage } from "../helper/$messageToIpcMessage.ts";
import { CommonEndpoint } from "./CommonEndpoint.ts";
import { ENDPOINT_PROTOCOL } from "./EndpointLifecycle.ts";
import type { $EndpointRawMessage } from "./EndpointMessage.ts";
export class WebMessageEndpoint extends CommonEndpoint {
  override toString(): string {
    return `WebMessageEndpoint#${this.debugId}`;
  }

  constructor(readonly port: MessagePort, debugId: string) {
    super(debugId);
    port.addEventListener("message", (event) => {
      const rawData = event.data;
      let message: $EndpointRawMessage;
      if (this.protocol === ENDPOINT_PROTOCOL.CBOR && typeof rawData !== "string") {
        message = $cborToEndpointMessage(rawData);
      } else {
        message = $jsonToEndpointMessage(rawData);
      }

      this.endpointMsgChannel.send(message);
    });
    port.start();
  }

  override doStart() {
    this.port.start();
    return super.doStart();
  }

  protected postTextMessage(data: String) {
    this.port.postMessage(data);
  }
  protected postBinaryMessage(data: Uint8Array) {
    this.port.postMessage(data);
  }
  protected override beforeClose = () => {};
}
